components/gitpod-db/go/encryption.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License.AGPL.txt in the project root for license information.
+
+package db
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+)
+
+type Cipher interface {
+	Encrypt(data []byte) (EncryptedData, error)
+	Decrypt(data EncryptedData) ([]byte, error)
+}
+
+func NewAES256CBCCipher(secret string, metadata CipherMetadata) (*AES256CBC, error) {
+	block, err := aes.NewCipher([]byte(secret))
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize AES 256 CBC cipher block: %w", err)
+	}
+
+	return &AES256CBC{
+		block:    block,
+		metadata: metadata,
+	}, nil
+}
+
+type AES256CBC struct {
+	block    cipher.Block
+	metadata CipherMetadata
+}
+
+func (c *AES256CBC) Encrypt(data []byte) (EncryptedData, error) {
+	iv, err := generateInitializationVector(16)
+	if err != nil {
+		return EncryptedData{}, err
+	}
+
+	// In CBC mode, the plaintext has to be padded to align with the cipher's block size
+	plaintext := pad(data, c.block.BlockSize())
+	ciphertext := make([]byte, len(plaintext))
+
+	cbc := cipher.NewCBCEncrypter(c.block, iv)
+	cbc.CryptBlocks(ciphertext, plaintext)
+
+	encoded := base64.RawStdEncoding.EncodeToString(ciphertext)
+	return EncryptedData{
+		EncodedData: encoded,
+		Params: KeyParams{
+			InitializationVector: string(iv),
+		},
+		Metadata: c.metadata,
+	}, nil
+}
+
+func (c *AES256CBC) Decrypt(data EncryptedData) ([]byte, error) {
+	if data.Metadata != c.metadata {
+		return nil, errors.New("cipher metadata does not match")
+	}
+
+	if data.Params.InitializationVector == "" {
+		return nil, errors.New("encrypted data does not contain an initialization vector")
+	}
+
+	ciphertext, err := data.Decode()
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode ciphertext: %w", err)
+	}
+
+	plaintext := make([]byte, len(ciphertext))
+
+	cbc := cipher.NewCBCDecrypter(c.block, []byte(data.Params.InitializationVector))
+	cbc.CryptBlocks(plaintext, ciphertext)
+
+	// In CBC mode, the plaintext was padded to align with the cipher's block size, we need to trim
+	trimmed := trim(plaintext)
+	return trimmed, nil
+}
+
+type KeyParams struct {
+	InitializationVector string `json:"iv"`
+}
+
+type CipherMetadata struct {
+	Name    string `json:"name"`
+	Version int    `json:"version"`
+}
+
+type EncryptedData struct {
+	// Data is base64 encoded
+	EncodedData string `json:"data"`
```

Comment:
Would be great to mention _why_ the JSON names are the way they are (backwards compatibility) to the existing impl.